### PR TITLE
feat(amplify-xcode): generate JSON schema

### PR DESCRIPTION
**Pull request by diegocstn**
*Description of changes:*
This PR introduces a new `amplify-xcode genenerate-schema` command to generate a JSON representation of the CLI and its commands as an intermediate step to further generate necessary JavaScript/TypeScript bindings.
In order to properly generate bindings and documentation for each commands, an interface describing parameters and their usage is generated according to the following specs:
```JS
{
  name: String, // parameter name
  help: String, // brief description
  kind: "option" | "flag" | "argument", // type of parameter (--option=value, --flag, <argument>),
  type: String, // string representation of parameter type, used to generate docs
}
```

Based on the current set of commands defined in `amplify-xcode`, the generated JSON schema will look like
```json
{
  "commands":[
     {
        "name":"import-config",
        "abstract":"Import Amplify configuration files",
        "parameters":[{ "kind":"option", "name":"path", "type":"String", "help":"Project base path"}]
     },
     {
        "name":"import-models",
        "abstract":"Import Amplify models",
        "parameters":[{ "kind":"option", "name":"path", "type":"String", "help":"Project base path"}]
     }
  ],
  "abstract":"Auto generated JSON representation of amplify-xcode CLI"
}
```
The following implementation is the results of a set of constraints/limitations in the ArgumentParser library and Swift runtime.

Using reflection to inspect a command type and generate a schema isn't a viable option due to the following constraints:
- ArgumentParser derives the name used on the command line from the name of the properties annotated with `@Option` `@Flag` property wrappers by using a private `String` extension, so for example a property named `fileOutputPath` will be converted to a `file-output-path` parameter when used to invoke the CLI
- trying to replicate the criteria used to derive parameters names may be inconsistent (how to handle special cases, like numbers and acronyms)
- both `@Option` and `@Flag` are generics on their wrapped value, that makes casting and/or infer their types using reflection almost impossible
- structs/classes used internally by `ArgumentParser` to parse value and/or generate help messages are private to the library and not exposed to its consumers
- accessing the enclosing type from a property wrappers isn't "officially" supported yet

*Check points:*

- [x] Added new tests to cover change, if needed
- [x] All unit tests pass
- [x] All integration tests pass
- [x] Documentation update for the change if required
- [x] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by palpatim** [comment]
General question: Is there an opportunity to use an existing schema definition such as [JSON Schema](https://json-schema.org/), [Swagger](https://swagger.io/), or [Smithy](https://awslabs.github.io/smithy/)? These are generally for service-side APIs, and I haven't investigated their applicability for a CLI use case, but it would be great if we could re-use prior art rather than creating [a new standard](https://xkcd.com/927/). :D

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodable.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+private enum CLICommandEncodableKeys: String, CodingKey {
+    case commandName
+    case abstract
+    case parameters
+}
+
+protocol CLICommandEncodable: Encodable {
+    static var commandName: String { get }
+    static var abstract: String { get }
+    static var paramsRegistry: CLICommandEncodableParametersRegistry { get }
```
Comment (palpatim) [inline AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodable.swift:20]: Do we need an additional wrapping type here, or can we just make `params` a `Set` with direct `get` access?

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommandEncodable.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    init(wrappedValue: Value, name: String, help: String, _ registry: CLICommandEncodableParametersRegistry) {
+        self.init(
+            wrappedValue: wrappedValue,
+            name: .customLong(name),
+            parsing: .next,
+            completion: nil,
+            help: ArgumentHelp(help)
+          )
+        let type = String(describing: Value.self)
+        registry.register(param: .option(name: name, type: type, help: help))
```
Comment (palpatim) [inline AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommandEncodable.swift:26]: As above, I wonder if the registry is necessary or whether we can just make `params` a `Set`

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodable.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import Foundation
+import ArgumentParser
+
+private enum CLICommandEncodableKeys: String, CodingKey {
+    case commandName
+    case abstract
+    case parameters
+}
+
+protocol CLICommandEncodable: Encodable {
```
[the remark this answers is not in the document]
Comment (diegocstn) [reply AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodable.swift:17]: there's no existing `CLICommand`, the only reason for the `CLICommandEncodable`'s existence is to have `Encodable` conformance -- but I'm thinking that would make sense to just have a generic `CLICommand` protocol and let it conform to `Encodable`. Thoughts?

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodable.swift (hunk 1 above)
Comment (diegocstn) [reply AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodable.swift:20]: We actually don't need it, I just wanted to add a layer of abstraction and hide the underlying storage, `Set`. (I don't even really like the name :P )

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodable.swift [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    static var commandName: String { get }
+    static var abstract: String { get }
+    static var paramsRegistry: CLICommandEncodableParametersRegistry { get }
+    init()
```
[the remark this answers is not in the document]
Comment (diegocstn) [reply AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodable.swift:21]: Good call, the only reason to have a zero-arg init is to be able to initialize a command and get its parameters set populated while we generate the schema. I'll define a separate `CLICommandInitializable`

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodableParameter.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// Encodable representation of CLI parameter.
+enum CLICommandEncodableParameter: Hashable {
```
[the remark this answers is not in the document]
Comment (diegocstn) [reply AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandEncodableParameter.swift:11]: Unfortunately there's no existing `CLICommandParameter` type as the parameters are being handled by the ArgumentParser library via property wrappers (`@Option`, `@Flag`) -- the only reason for having an extra type is to retain information about commands' parameters and use it for generating the schema

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandGenerateJSONSchema.swift [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import Foundation
+import AmplifyXcodeCore
+import ArgumentParser
+
+private struct CLISchema: Encodable {
+    let abstract = "Auto generated JSON representation of amplify-xcode CLI"
+    var commands: [AnyCLICommandEncodable] = []
+
+    init() {
+        for command in AmplifyXcode.configuration.subcommands where command != CLICommandGenerateJSONSchema.self {
```
[the remark this answers is not in the document]
Comment (brennanMKE) [reply AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandGenerateJSONSchema.swift:17]: +1

[the remark this answers is not in the document]
Comment (diegocstn) [reply AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandGenerateJSONSchema.swift]: Updated, `generate-schema` command generates a schema of itself :)

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommand.swift [hunk 7]
```diff
@@ -11,10 +11,11 @@ import ArgumentParser
 /// The following extensions on ArgumentParser commands parameters property wrappers
 /// help us providing a hook to generate a JSON representation of a CLI command.
 /// As for now `@PropertyWrappers` APIs to access enclosing type are still "private", so the passed
-/// `registry` allows the property to reference an external type.
+/// `parameters` allows the property to reference an external type.
 /// `Argument` has been left out on purpose as we'd rather use options and flags for clarity of use.
+/// Also by providing these extra initializers we make parameter name explicit.
 extension Option where Value: ExpressibleByArgument {
-    init(wrappedValue: Value, name: String, help: String, _ registry: CLICommandEncodableParametersRegistry) {
+    init(wrappedValue: Value, name: String, help: String, _ parameters: inout Set<CLICommandParameter>) {
```
Comment (palpatim) [inline AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommand.swift:18]: Minor naming suggestion to clarify the intent of the `parameters` and why it's an `inout`:
```suggestion
    init(wrappedValue: Value, name: String, help: String, updating parameters: inout Set<CLICommandParameter>) {
```

Result at the call site would look like:
```swift
@Option(name: "path", help: "Project base path", updating: &parameters)
...
```

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandImportConfig.swift [hunk 8]
```diff
@@ -10,14 +10,14 @@ import ArgumentParser
 import AmplifyXcodeCore
 
 /// CLI command invoking `CommandImportConfig`.
-struct CLICommandImportConfig: ParsableCommand, CommandExecutable, CLICommandReportable, CLICommandEncodable {
-    static var paramsRegistry = CLICommandEncodableParametersRegistry()
+struct CLICommandImportConfig: ParsableCommand, CommandExecutable, CLICommandReportable, CLICommand {
+    static var parameters: Set<CLICommandParameter> = []
```
Comment (palpatim) [inline AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandImportConfig.swift:14]: Style nit: We've been preferring to let type inference do its thing:
```suggestion
    static var parameters = Set<CLICommandParameter>()
```

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandImportModels.swift [hunk 9]
```diff
@@ -10,14 +10,14 @@ import ArgumentParser
 import AmplifyXcodeCore
 
 /// CLI command invoking `CommandImportModels`.
-struct CLICommandImportModels: ParsableCommand, CommandExecutable, CLICommandReportable, CLICommandEncodable {
-    static var paramsRegistry = CLICommandEncodableParametersRegistry()
+struct CLICommandImportModels: ParsableCommand, CommandExecutable, CLICommandReportable, CLICommand {
+    static var parameters: Set<CLICommandParameter> = []
```
Comment (palpatim) [inline AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandImportModels.swift:14]: nit
```suggestion
    static var parameters = Set<CLICommandParameter>()
```

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandParameter.swift [hunk 10]
```diff
@@ -7,8 +7,13 @@
...
 import Foundation
 
-/// Encodable representation of CLI parameter.
-enum CLICommandEncodableParameter: Hashable {
+/// Encodable representation of a CLI command parameter.
+/// Commands parameters (options, flags, arguments) are declared as properties on the command type
+/// and annotated with `@propertyWrapper`s `@Option`, `@Flag` and `@Argument` provided by `ArgumentParser`.
+/// `ArgumentParser` derives parameters names from property names (i.e.,  an`outputPath` option becomes `--output-path`)
+/// making thus impossible to reliably generate a JSON representation of a command and its parameters.
+/// Therefore we use the following enum to keep track of each parameter and their attributes (name, type and help text).
```
Comment (palpatim) [inline AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/CLICommandParameter.swift:15]: :+1:

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommand.swift [hunk 11]
```diff
@@ -23,25 +24,25 @@ extension Option where Value: ExpressibleByArgument {
             help: ArgumentHelp(help)
           )
         let type = String(describing: Value.self)
-        registry.register(param: .option(name: name, type: type, help: help))
+        parameters.insert(.option(name: name, type: type, help: help))
     }
 
-    init(name: String, help: String, _ registry: CLICommandEncodableParametersRegistry) {
+    init(name: String, help: String, _ parameters: inout Set<CLICommandParameter>) {
```
Comment (palpatim) [inline AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommand.swift:30]:
```suggestion
    init(name: String, help: String, updating parameters: inout Set<CLICommandParameter>) {
```

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommand.swift [hunk 12]
```diff
@@ -23,25 +24,25 @@ extension Option where Value: ExpressibleByArgument {
...
           )
         let type = String(describing: Value.self)
-        registry.register(param: .option(name: name, type: type, help: help))
+        parameters.insert(.option(name: name, type: type, help: help))
     }
 }
 
 extension Flag where Value == Bool {
-    init(wrappedValue: Value, name: String, help: String, _ registry: CLICommandEncodableParametersRegistry) {
+    init(wrappedValue: Value, name: String, help: String, _ parameters: inout Set<CLICommandParameter>) {
```
Comment (palpatim) [inline AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommand.swift:43]:
```suggestion
    init(wrappedValue: Value, name: String, help: String, updating parameters: inout Set<CLICommandParameter>) {
```

**Review by palpatim — APPROVED** [review]
Approved w/comments

File: AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommand.swift (hunk 7 above)
Comment (diegocstn) [reply AmplifyTools/AmplifyXcode/Sources/AmplifyXcode/Support/ArgumentParser+CLICommand.swift:18]: good call, thanks!

**Pull request merged**